numbat/src/interpreter.rs

```diff
@@ -137,137 +141,112 @@ mod tests {
-    }
-
-    #[track_caller]
-    fn assert_runtime_error(input: &str, err_expected: RuntimeError) {
-        if let Err(err_actual) = get_interpreter_result(input) {
-            assert_eq!(err_actual, err_expected);
-        } else {
-            panic!();
+    fn evaluates(input: impl AsRef<str>) -> String {
+        match get_interpreter_result(input.as_ref()) {
+            Ok(InterpreterResult::Value(value)) => format!("{}", value),
+            Ok(_) => panic!("Received non value result"),
+            Err(e) => format!("{e:?}: `{e}`"),
         }
     }
 
     #[test]
     fn simple_arithmetic() {
-        assert_evaluates_to_scalar("0", 0.0);
-        assert_evaluates_to_scalar("1", 1.0);
-        assert_evaluates_to_scalar("1+2", 1.0 + 2.0);
-        assert_evaluates_to_scalar("-1", -1.0);
-
-        assert_evaluates_to_scalar("2+3*4", 2.0 + 3.0 * 4.0);
-        assert_evaluates_to_scalar("2*3+4", 2.0 * 3.0 + 4.0);
-        assert_evaluates_to_scalar("(2+3)*4", (2.0 + 3.0) * 4.0);
-
-        assert_evaluates_to_scalar("(2/3)*4", (2.0 / 3.0) * 4.0);
-        assert_evaluates_to_scalar("-2 * 3", -2.0 * 3.0);
-        assert_evaluates_to_scalar("2 * -3", 2.0 * -3.0);
-        assert_evaluates_to_scalar("2 - 3 - 4", 2.0 - 3.0 - 4.0);
-        assert_evaluates_to_scalar("2 - -3", 2.0 - -3.0);
-
-        assert_evaluates_to_scalar("+2 * 3", 2.0 * 3.0);
-        assert_evaluates_to_scalar("2 * +3", 2.0 * 3.0);
-        assert_evaluates_to_scalar("+2 - +3", 2.0 - 3.0);
+        assert_snapshot!(evaluates("0"), @"0");
+        assert_snapshot!(evaluates("1"), @"1");
+        assert_snapshot!(evaluates("1+2"), @"3");
+        assert_snapshot!(evaluates("-1"), @"-1");
+
+        assert_snapshot!(evaluates("2+3*4"), @"14");
+        assert_snapshot!(evaluates("2*3+4"), @"10");
+        assert_snapshot!(evaluates("(2+3)*4"), @"20");
+
+        assert_snapshot!(evaluates("(2/3)*4"), @"2.66667");
+        assert_snapshot!(evaluates("-2 * 3"), @"-6");
+        assert_snapshot!(evaluates("2 * -3"), @"-6");
+        assert_snapshot!(evaluates("2 - 3 - 4"), @"-5");
+        assert_snapshot!(evaluates("2 - -3"), @"5");
+
+        assert_snapshot!(evaluates("+2 * 3"), @"6");
+        assert_snapshot!(evaluates("2 * +3"), @"6");
+        assert_snapshot!(evaluates("+2 - +3"), @"-1");
     }
 
     #[test]
     fn arithmetic_with_units() {
-        use crate::unit::Unit;
-
-        assert_evaluates_to(
-            "2 meter + 3 meter",
-            Quantity::from_scalar(2.0 + 3.0) * Quantity::from_unit(Unit::meter()),
-        );
+        assert_snapshot!(evaluates(
+            "2 meter + 3 meter"), @"5 m");
 
-        assert_evaluates_to(
+        assert_snapshot!(evaluates(
             "dimension Pixel
              @aliases(px: short)
              unit pixel : Pixel
-             2 * pixel",
-            Quantity::from_scalar(2.0) * Quantity::from_unit(Unit::new_base("pixel", "px")),
-        );
+             2 * pixel"), @"2 px");
```

Comment:
This is shorter and looks nicer than what we had, but it's also not quite the same. Since you're using `format!` for comparison here, these tests  (this one and below) now depend on the `Display` implementations of `Quantity`/`Number`/`Unit`, which are not-trivial. One problem with this is that we might have to update the tests frequently when we change those Display impls (I understand that this is much more convenient now with `insta`). The other problem is that this might hide things (think of rounding, for example).

What do you think?

Reply:
Yeah, it’s what I wanted to see with you.
On one side, it hides information, on the other, don't we want to notice when the final message changes?

IMO, the other options are:
- Use the `Debug` implem
- Display the `Display` + `Debug` implem side by side (`format!("{} => {:?}")`) just to be sure we don't miss anything

Reply:
> Display the `Display` + `Debug` implem side by side (`format!("{} => {:?}")`) just to be sure we don't miss anything

I like this idea. If the output is not too noisy, this might solve all issues.

Reply:
Hey, once again, I’m late 😅 

What do you think of this last commit? 
Do you think it's too noisy, or are we good, and I can start applying the same techniques to more files?